crypto/sessionkey.go

```diff
@@ -0,0 +1,211 @@
+	constants.CAST5:     packet.CipherCAST5,
+	constants.AES128:    packet.CipherAES128,
+	constants.AES192:    packet.CipherAES192,
+	constants.AES256:    packet.CipherAES256,
+}
+
+// GetCipherFunc returns the cipher function corresponding to the algorithm used
+// with this SessionKey.
+func (sk *SessionKey) GetCipherFunc() packet.CipherFunction {
+	cf, ok := symKeyAlgos[sk.Algo]
+	if ok {
+		return cf
+	}
+
+	panic("gopenpgp: unsupported cipher function: " + sk.Algo)
+}
+
+// GetBase64Key returns the session key as base64 encoded string.
+func (sk *SessionKey) GetBase64Key() string {
+	return base64.StdEncoding.EncodeToString(sk.Key)
+}
+
+// RandomToken generates a random token with the specified key size
+func RandomToken(size int) ([]byte, error) {
+	config := &packet.Config{DefaultCipher: packet.CipherAES256}
+	symKey := make([]byte, size)
+	if _, err := io.ReadFull(config.Random(), symKey); err != nil {
+		return nil, err
+	}
+	return symKey, nil
+}
+
+// GenerateSessionKeyAlgo generates a random key of the correct length for the specified algorithm
+func GenerateSessionKeyAlgo(algo string) (sk *SessionKey, err error) {
+	cf, ok := symKeyAlgos[algo]
+	if !ok {
+		return nil, errors.New("gopenpgp: unknown symmetric key generation algorithm")
+	}
+	r, err := RandomToken(cf.KeySize())
+	if err != nil {
+		return nil, err
+	}
+
+	sk = &SessionKey{
+		Key: r,
+		Algo: algo,
+	}
+	return sk, nil
+}
+
+// GenerateSessionKey generates a random key for the default cipher
+func GenerateSessionKey() (*SessionKey, error) {
+	return GenerateSessionKeyAlgo(constants.AES256)
+}
+
+func NewSessionKeyFromToken(token []byte, algo string) *SessionKey {
+	return &SessionKey{
+		Key:  token,
+		Algo: algo,
+	}
+}
+
+func newSessionKeyFromEncrypted(ek *packet.EncryptedKey) (*SessionKey, error) {
+	var algo string
+	for k, v := range symKeyAlgos {
+		if v == ek.CipherFunc {
+			algo = k
+			break
+		}
+	}
+	if algo == "" {
+		return nil, fmt.Errorf("gopenpgp: unsupported cipher function: %v", ek.CipherFunc)
+	}
+
+	symmetricKey := &SessionKey{
+		Key:  ek.Key,
+		Algo: algo,
+	}
+	
+	return symmetricKey, nil
+}
+
+// Encrypt encrypts a PlainMessage to PGPMessage with a SessionKey
+// * message : The plain data as a PlainMessage
+// * output  : The encrypted data as PGPMessage
+func (sk *SessionKey) Encrypt(message *PlainMessage) ([]byte, error) {
+	var encBuf bytes.Buffer
+	var encryptWriter io.WriteCloser
+	config := &packet.Config{
+		Time:          getTimeGenerator(),
+		DefaultCipher: sk.GetCipherFunc(),
+	}
+
+	encryptWriter, err := packet.SerializeSymmetricallyEncrypted(&encBuf, config.Cipher(), sk.Key, config)
+	if err != nil {
+		return nil, errors.Wrap(err, "gopenpgp: unable to encrypt")
+	}
+
+	if algo := config.Compression(); algo != packet.CompressionNone {
+		encryptWriter, err = packet.SerializeCompressed(encryptWriter, algo, config.CompressionConfig)
+		if err != nil {
+			return nil, errors.Wrap(err, "gopenpgp: unable to encrypt")
```

Comment:
There is already `gopenpgp: unable to encrypt` few lines above. Would be good to have different message so it can be easier to spot from where the error is coming from. Also, I have feeling this is more `unable to serialize` than encrypt, anyway?

Reply:
`unable to serialize` is already a few lines under where we actually pass the data to serialize, i will go for `error in compression` :)